src/pool.jl

```diff
@@ -3,39 +3,46 @@ function CategoricalPool(index::Vector{S},
                          order::Vector{R},
                          ordered::Bool=false) where {S, T <: Integer, R <: Integer}
     invindex = convert(Dict{S, R}, invindex)
-    CategoricalPool{S, R, CategoricalValue{S, R}}(index, invindex, order, ordered)
+    C = catvalue_type(S, R)
+    V = valtype(C) # might be diffrent from S (e.g. S==SubString, V==String)
+    CategoricalPool{V, R, C}(index, invindex, order, ordered)
 end
 
+CategoricalPool{T, R, C}(ordered::Bool=false) where {T, R, C} =
+    CategoricalPool{T, R, C}(T[], Dict{T, R}(), R[], ordered)
 CategoricalPool{T, R}(ordered::Bool=false) where {T, R} =
     CategoricalPool(T[], Dict{T, R}(), R[], ordered)
 CategoricalPool{T}(ordered::Bool=false) where {T} =
-    CategoricalPool(T[], Dict{T, DefaultRefType}(), DefaultRefType[], ordered)
+    CategoricalPool{T, reftype(T)}(ordered)
```

Comment:
Do we really need to call `reftype` here and below? I would assume `T` cannot be a categorical value, as `CategoricalArray` constructors should have controlled the parameters already?

Reply:
`reftype(T)` returns the default reference type for non-"categorical value" types. I.e. `DefaultRefType`.
It could be extended to return e.g. `UInt8` for `UInt8` or from the user code for custom types.

Some of these ctors are used in `recode.jl` or `extras.jl` (outside of `CategoricalArray`), some are just tested.
I can look into removing `CategoricalPool` ctor that are only used from tests.
I will add the check that `V` cannot be a categorical value type into the inner ctor of `CategoricalPool` (plus `reftype(C) === R` and `valtype(C) === V` checks).

Reply:
> reftype(T) returns the default reference type for non-"categorical value" types. I.e. DefaultRefType.
> It could be extended to return e.g. UInt8 for UInt8 or from the user code for custom types.

As I noted before, please implement the simplest system for now and leave possible extensions by external packages for later. So here let's just keep `DefaultRefType`, the diff is already large enough.

Indeed these constructors existed before `CategoricalArray`, so some of them may not be useful and could be removed in another PR.

Reply:
I have considered reverting `reftype(T)` to `DefaultRefType`, but then realized it would result in resetting of reference type in constructors like `CategoricalArray(x::AbstractArray)`, where `x` could actually be `CategoricalArray`.
I don't think it would be the expected behaviour, so I'll keep the generic `reftype(T)`.

Reply:
But we're dealing with the `CategoricalPool` constructor here, not the `CategoricalArray` one. If `T` is a categorical value, we're screwed anyway since there's no code in the constructors to unwrap `T` -- and indeed you just added checks to throw an error in that situation.